sympy/matrices/matrices.py

```diff
@@ -1581,6 +1583,12 @@ def _matrix_pow_by_recursion(self, num):
     def __pow__(self, num):
         if not self.is_square:
             raise NonSquareMatrixError()
+        num = sympify(num)
+        if not num.is_integer and (num >= 0) == True and not num % 1 == 0:
```

Comment:
In sympy\physics\quantum\density.py
```
>>> fidelity(up_dm, down_dm) #orthogonal states
    0
```
I think this is going to fail under the new conditions (has nilpotent jordan block with size > 1)